crates/oxc_transformer/src/regexp/mod.rs

```diff
@@ -0,0 +1,237 @@
+//!
+//! ### ES2015
+//!
+//! #### Sticky flag (`y`)
+//! - @babel/plugin-transform-sticky-regex: <https://babeljs.io/docs/en/babel-plugin-transform-sticky-regex>
+//!
+//! #### Unicode flag (`u`)
+//! - @babel/plugin-transform-unicode-regex: <https://babeljs.io/docs/en/babel-plugin-transform-unicode-regex>
+//!
+//! ### ES2018
+//!
+//! #### DotAll flag (`s`)
+//! - @babel/plugin-transform-dotall-regex: <https://babeljs.io/docs/en/babel-plugin-transform-dotall-regex>
+//! - Spec: ECMAScript 2018: <https://262.ecma-international.org/9.0/#sec-get-regexp.prototype.dotAll>
+//!
+//! #### Lookbehind assertions (`/(?<=x)/` and `/(?<!x)/`)
+//! - Implementation: Same as esbuild's handling
+//!
+//! #### Named capture groups (`(?<name>x)`)
+//! - @babel/plugin-transform-named-capturing-groups-regex: <https://babeljs.io/docs/en/babel-plugin-transform-named-capturing-groups-regex>
+//!
+//! #### Unicode property escapes (`\p{...}` and `\P{...}`)
+//! - @babel/plugin-transform-unicode-property-regex: <https://babeljs.io/docs/en/babel-plugin-proposal-unicode-property-regex>
+//!
+//! ### ES2022
+//!
+//! #### Match indices flag (`d`)
+//! - Implementation: Same as esbuild's handling
+//!
+//! ### ES2024
+//!
+//! #### Set notation + properties of strings (`v`)
+//! - @babel/plugin-transform-unicode-sets-regex: <https://babeljs.io/docs/en/babel-plugin-proposal-unicode-sets-regex>
+//! - TC39 Proposal: <https://github.com/tc39/proposal-regexp-set-notation>
+
+mod options;
+
+use std::borrow::Cow;
+
+pub use options::RegExpOptions;
+use oxc_ast::ast::*;
+use oxc_regular_expression::ast::{LookAroundAssertionKind, Pattern, Term};
+use oxc_semantic::ReferenceFlags;
+use oxc_traverse::Traverse;
+
+use crate::context::Ctx;
+
+pub struct RegExp<'a> {
+    ctx: Ctx<'a>,
+    options: RegExpOptions,
+}
+
+impl<'a> RegExp<'a> {
+    pub fn new(options: RegExpOptions, ctx: Ctx<'a>) -> Self {
+        Self { ctx, options }
+    }
+}
+
+impl<'a> Traverse<'a> for RegExp<'a> {
+    fn enter_expression(
+        &mut self,
+        expr: &mut Expression<'a>,
+        ctx: &mut oxc_traverse::TraverseCtx<'a>,
+    ) {
+        let Expression::RegExpLiteral(regexp) = expr else {
+            return;
+        };
+
+        let is_unsupported = self.has_unsupported_regular_expression_flags(regexp.regex.flags);
+
+        let mut pattern_source = Cow::Borrowed(match regexp.regex.pattern {
+            RegExpPattern::Raw(pattern)
+                if is_unsupported
+                    || self.has_unsupported_regex_syntax_raw(pattern, regexp.regex.flags) =>
+            {
+                pattern
+            }
+            RegExpPattern::Pattern(ref pattern)
+                if is_unsupported || self.has_unsupported_regular_expression_pattern(pattern) =>
+            {
+                regexp.regex.pattern.source_text(self.ctx.source_text)
+            }
+            _ => return,
+        });
```

Comment:
Now I need to handle both `RegExpPattern::Raw` and `RegExpPattern::Pattern`, it's not ideal 🥲

Reply:
I'm not familiar with how the new regex parser works. How do we end up with `RegExpPattern::Raw`? Is parsing regexes optional?

Reply:
@rzvxa What the difference between `Raw` and `Invalid`?

Reply:
> I'm not familiar with how the new regex parser works. How do we end up with `RegExpPattern::Raw`? Is parsing regexes optional?

Yes, The parser has a `parse_regular_expression` option to enable it.

https://github.com/oxc-project/oxc/blob/abb1e6f2ce3fa4f28422febe76cde674be43c7ab/crates/oxc_parser/src/js/expression.rs#L347-L357

So we have to check both

Reply:
> @rzvxa What was the difference between `Raw` and `Invalid`?

`Raw` is when the regex parser is off, and `Invalid` is when the regex string is invalid and we fail to parse it.

They basically contain the same piece of information just tagged differently. You can use the `RegExpPattern::as_pattern` method to get it as an option when you don't care to differentiate between `Raw` and `Invalid`.

Reply:
> > I'm not familiar with how the new regex parser works. How do we end up with `RegExpPattern::Raw`? Is parsing regexes optional?
> 
> Yes, The parser has a `parse_regular_expression` option to enable it.
> 
> https://github.com/oxc-project/oxc/blob/abb1e6f2ce3fa4f28422febe76cde674be43c7ab/crates/oxc_parser/src/js/expression.rs#L347-L357
> 
> So we have to check both

We can add a helper method to the `RegExpLiteral` that would parse the pattern. Something like this:

```rust
impl<'a> RegExpLiteral<'a> {
    pub fn try_parse_pattern(
        &self,
        allocator: &'a Allocator,
    ) -> Result<Pattern<'a>, OxcDiagnostic> {
        match &self.regex.pattern {
            RegExpPattern::Raw(raw) | RegExpPattern::Invalid(raw) => {
                use oxc_regular_expression::{ParserOptions, PatternParser};
                let options = ParserOptions {
                    span_offset: self.span.start + 1, // exclude `/`
                    unicode_mode: self.regex.flags.contains(RegExpFlags::U)
                        || self.regex.flags.contains(RegExpFlags::V),
                    unicode_sets_mode: self.regex.flags.contains(RegExpFlags::V),
                };
                PatternParser::new(allocator, raw, options).parse()
            }
            RegExpPattern::Pattern(it) => Ok(it.as_ref().clone_in(allocator)),
        }
    }
}
```

If we add a span field to the `RegExp` type this method can exist on that type instead. At the moment `RegExpLiteral` is the closest place to `RegExpPattern` which contains a concrete span.

Reply:
I just want to add that we can build upon my previous suggestion and add a lazy parsing that would take mutable reference to the AST node and parse it in place whenever we need to access it.

Reply:
> We can add a helper method to the `RegExpLiteral` that would parse the pattern. Something like this:
> 
> ```rust
> impl<'a> RegExpLiteral<'a> {
>     pub fn try_parse_pattern(
>         &self,
>         allocator: &'a Allocator,
>     ) -> Result<Pattern<'a>, OxcDiagnostic> { // It should return a `CowLike<Pattern>` to avoid redundant cloning when it is already parsed.
>         match &self.regex.pattern {
>             // Parse even if pattern is known to be invalid so we can reproduce the errors for the caller.
>             RegExpPattern::Raw(raw) | RegExpPattern::Invalid(raw) => {
>                 use oxc_regular_expression::{ParserOptions, PatternParser};
>                 let options = ParserOptions {
>                     span_offset: self.span.start + 1, // exclude `/`
>                     unicode_mode: self.regex.flags.contains(RegExpFlags::U)
>                         || self.regex.flags.contains(RegExpFlags::V),
>                     unicode_sets_mode: self.regex.flags.contains(RegExpFlags::V),
>                 };
>                 PatternParser::new(allocator, raw, options).parse()
>             }
>             RegExpPattern::Pattern(it) => Ok(it.as_ref().clone_in(allocator)),
>         }
>     }
> }
> ```
> 
> If we add a span field to the `RegExp` type this method can exist on that type instead. At the moment `RegExpLiteral` is the closest place to `RegExpPattern` which contains a concrete span.

Actually, Here just need to check whether the pattern contains a kind of syntax. Full parsing of regexp is too costly